File: dynamic_programming/python/309_best_time_to_buy_and_sell_stock_with_cooldown.py
#define MAX(x,y) (((x) > (y)) ? (x) : (y))
#define MIN(x,y) (((x) < (y)) ? (x) : (y))

class solution:
    def maxProfit(prices:list[int])->int:
        pricesSize = len(prices)
        
        dp_sell_out = []
        dp_sell_with = []
        dp_buy = []
        
        #current action is not selling, next step could be buying
        dp_sell_out.append(0)
        # current action is selling, next step must be cooldown
        dp_sell_with.append(0)

        # make sure last action is buying
        dp_buy.append(-prices[0])

        for i in range(1,pricesSize):
            dp_sell_out.append(max(dp_sell_out[i-1],dp_sell_with[i-1]))
            dp_sell_with.append(dp_buy[i-1]+prices[i])
            dp_buy.append(max(dp_buy[i-1],dp_sell_out[i-1]-prices[i]))

        return max(dp_sell_out[pricesSize-1],dp_sell_with[pricesSize-1])

File: dynamic_programming/c++/309_best_time_to_buy_and_sell_stock_with_cooldown.cpp
#define MAX(x,y) (((x) > (y)) ? (x) : (y))
#define MIN(x,y) (((x) < (y)) ? (x) : (y))

class solution{
public:
    int maxProfit(vector<int>&prices){
        int pricesSize = prices.size();
        int dp_sell_out[pricesSize];
        int dp_sell_with[pricesSize];
        int dp_buy[pricesSize];

        //not bought any stock at current step 
        dp_sell_out[0] = 0;

        //bought stock at current step
        dp_sell_with[0] = 0;
        dp_buy[0] = -prices[0];

        for(int i=1;i<pricesSize;i++){
            dp_sell_out[i] = max(dp_sell_out[i-1],dp_sell_with[i-1]); 
            dp_sell_with[i] = dp_buy[i-1]+prices[i];
            dp_buy[i] = max(dp_buy[i-1], dp_sell_out[i-1]-prices[i]);
        }
        return max(dp_sell_out[pricesSize-1],dp_sell_with[pricesSize-1]);    
    }
};
